rls: update picker synchronously on configuration update

aranjans opened the pull request:
fixes #7212.

### Problem
#5469 recommends an audit of existing LB policies to ensure that they update their pickers synchronously upon receipt of a configuration update. RLS lb policy do updates the picker synchronously but misses out in one single corner case when there is a config change in cache size, where we need to send picker update in case any cache entry with valid backoff was evicted.

### What does this PR do?
This PR handles that corner cases and make sure it updates the picker in all of the cases synchronously with config update of RLS lb policy.

RELEASE NOTES:
- rls: update picker synchronously on configuration update.

arjan-bal commented on balancer/rls/balancer.go:
```diff
@@ -297,13 +298,19 @@ func (b *rlsBalancer) UpdateClientConnState(ccs balancer.ClientConnState) error
-		b.dataCache.resize(newCfg.cacheSizeBytes)
+		evicted := b.dataCache.resize(newCfg.cacheSizeBytes)
+		if evicted {
+			b.sendNewPickerLocked()
```
Since we are interested if testing that the picker was updated, we should put the hook inside `sendNewPickerLocked`. Presently, someone can delete the line `b.sendNewPickerLocked()` and the test would still pass as long as the call to `entryWithValidBackoffEvicted()` is there below.

arjan-bal reviewed (commented):
Great progress since the last attempt!

aranjans commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +650,201 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+	// these as targets in the RLS response, based on request keys.
+	_, backendAddress1 := startBackend(t)
+	backendCh2, backendAddress2 := startBackend(t)
+	backendCh3, backendAddress3 := startBackend(t)
```
We would need the third backend as well, bcz first one won't be successful RPC as control channel will throw error and put that in backoff. When we make RPC call to backend2 and backend3, it will be actually go into the cache and gets evicted on the config update with resizing of cache.

aranjans commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +650,201 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+	lis := testutils.NewRestartableListener(l)
+
+	// Override the clientConn update hook to get notified.
+	clientConnUpdateDone := make(chan struct{}, 1)
```
thAnks for the catch, you're right. We should check for the picker update before lb config update is complete which would make sure these udpates are synchronous.

aranjans commented:
@arjan-bal I have updated the PR and referenced your comments in the code change.

arjan-bal commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +650,183 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+
+	// Override the entryWithValidBackoffEvicted to ensure we update
+	// picker when an entry with valid backoff time was evicted.
+	backOffItemEvicted := make(chan struct{}, 1)
```
This needs to be a blocking channel if we want to test for synchronous updates, see the comment above.
```go
backOffItemEvicted := make(chan struct{})
```

arjan-bal commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +650,183 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+                "n2"
+            ]
+        },
+        {
+            "key": "k3",
+            "names": [
+                "n3"
+            ]
+        }
```
I believe this can be removed as there are only 2 backends now.

arjan-bal commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +650,183 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+		if req.KeyMap["k2"] == "v2" {
+			return &rlstest.RouteLookupResponse{Resp: &rlspb.RouteLookupResponse{Targets: []string{backendAddress1}}}
+		}
+		if req.KeyMap["k3"] == "v3" {
+			return &rlstest.RouteLookupResponse{Resp: &rlspb.RouteLookupResponse{Targets: []string{backendAddress2}}}
+		}
```
I believe this can be removed now.

arjan-bal commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +650,183 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+	makeTestRPCAndExpectItToReachBackend(ctxOutgoing, t, cc, backendCh1)
+	verifyRLSRequest(t, rlsReqCh, true)
+
+	ctxOutgoing = metadata.AppendToOutgoingContext(ctx, "n3", "v3")
+	makeTestRPCAndExpectItToReachBackend(ctxOutgoing, t, cc, backendCh2)
+	verifyRLSRequest(t, rlsReqCh, true)
```
Is this still needed?

aranjans replied:
yes,  we'd need this RPC call. I have verified it without having this RPC call, and the test fails.

arjan-bal replied:
I was referring to the call with metadata values `"n3": "v3"` which was expected to go to backend 3 before the changes. I see that its removed now.

arjan-bal commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +650,183 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+	entryWithValidBackoffEvicted = func() {
+		select {
+		case backOffItemEvicted <- struct{}{}:
+		default:
+			// Do nothing if the channel is full
```
We should block until the test reads from the channel. This is how we verify that the update doesn't complete before the picker is updated.

arjan-bal commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +650,183 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+	// synchronously when there was cache resize on config update.
+	case <-backOffItemEvicted:
+	case <-ctx.Done():
+		t.Errorf("Error sending picker update on eviction of cache entry with valid backoff: %v", ctx.Err().Error())
```
There should be another case for `<-clientConnUpdateDone` which should result in an error that `client conn update was completed before the picker update`.

aranjans commented:
@arjan-bal thAnks for your review, I have updated the PR with your suggestions. 
fyi: For all the suggestions, I have just put the thumbs up to the comment and updated the PR.

arjan-bal commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +650,181 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+	// Note: This needs to be initialised here and not in the beginning
+	// of the test, as otherwise for every RPC call we make, it'd have
+	// affected total picker count.
+	pckrSentBeforeClientConnUpdate := make(chan struct{}, 1)
+	totalPickerGenerated := 0
+	origNewPickerGenerated := newPickerGenerated
+	newPickerGenerated = func() {
+		totalPickerGenerated++
+		if totalPickerGenerated == 2 {
+			pckrSentBeforeClientConnUpdate <- struct{}{}
+		}
+	}
```
Instead of maintaining a counter and and a channel, this can be done using a blocking channel.
```go
pickerUpdateChan := make(chan struct{})
newPickerGenerated = func() {
	newPickerGenerated<-struct{}{}
}

// snipped
for i := 0; i < 2; i++{
	select {
	case <-pckrSentBeforeClientConnUpdate:
	case <-clientConnUpdateDone:
		t.Fatalf("Client conn update was completed before picker update.")
	case <-ctx.Done():
		t.Fatalf("client conn update could not complete: %v", ctx.Err().Error())
	}
}
```

aranjans replied:
Actually that might not be correct as we are making the RPC calls in between, so that could also make the picker updated. For ease of the test I think the current state of PR makes more sense to me, as it specifically verifies for status updates that there should be only 2 picker updates happening on cache resizing. WDYT?

arjan-bal replied:
Usually test setup happens before the test is run and results are verified. 
If you want to prevent the blocking till the last update, you can use an atomic boolean like `blockUpdates := atomic.Bool{}`.

In the picker update hook, you can check the bool and let the request through. Just before calling `UpdateState`, you can set the boolean to true.

arjan-bal replied:
Also reading from the blocking channel twice is a cleaner way than having the hook send one message on the channel after 2 calls.

arjan-bal commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +650,181 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+	makeTestRPCAndExpectItToReachBackend(ctxOutgoing, t, cc, backendCh2)
+	verifyRLSRequest(t, rlsReqCh, true)
+
+	// Override the newPickerGenerated to measure the number of times
+	// the picker is generated because state updates can be inhibited.
+	//
+	// Note: This needs to be initialised here and not in the beginning
+	// of the test, as otherwise for every RPC call we make, it'd have
+	// affected total picker count.
```
Even if we were using a counter, we could move this to the beginning and check for the difference in the counter values before and after calling `UpdateClientConnState`. 

As described in the comment below, I don't think we need a counter. We should move this to the top and verify the order of pickerGeneration above also.

arjan-bal commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +650,181 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+		t.Fatalf("Client conn update was completed before picker update.")
+	case <-ctx.Done():
+		t.Errorf("client conn update could not complete: %v", ctx.Err().Error())
+	}
```
We need to wait for `<-clientConnUpdateDone` at the end.

arjan-bal commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +650,180 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+	// Note: This needs to be initialised here and not in the beginning
+	// of the test, as otherwise for every RPC call we make, it'd have
+	// affected total picker count.
+	pckrSentBeforeClientConnUpdate := make(chan struct{}, 1)
```
This should be a blocking channel.

arjan-bal replied:
This wasn't addressed.

arjan-bal commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +650,180 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+
+	// Once picker was updated, wait for client conn update
+	// to complete.
+	<-clientConnUpdateDone
```
Its probably a good idea to wait for context timeout along with this to ensure the test doesn't hang for a long time in case of failures.

arjan-bal commented:
Please address comments before re-assigning to me :)

aranjans commented:
Ah my bad, chrome tab wasn't refreshed and i thoughts its still not reviewed.

arjan-bal reviewed (commented):
For future, please ensure comments made in one place are resolved throughout the PR when applicable. I feel that the similar changes are requested a couple times now. This is dragging on the review for longer than necessary.

aranjans commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +651,188 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+
+	// Override the newPickerGenerated to measure the number of times
+	// the picker is generated because state updates can be inhibited.
+	pckrSentBeforeClientConnUpdate := make(chan struct{}, 1)
```
Why do we need to wait for the picker update twice?
We are already blocking updates until the last update using atomic.Bool{}, so we will only have one struct going into `pckrSentBeforeClientConnUpdate`

arjan-bal commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +651,191 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+		case <-clientConnUpdateDone:
+			t.Fatalf("Client conn update was completed before picker update.")
+		case <-ctx.Done():
+			t.Errorf("error waiting for picker update on receipt of configuration udpate: %v", ctx.Err().Error())
```
nit: Capital E in "Error waiting.."

easwars replied:
Again thumbs up without any changes.

arjan-bal commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +651,191 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+	select {
+	case <-clientConnUpdateDone:
+	case <-ctx.Done():
+		t.Errorf("client conn update could not complete: %v", ctx.Err().Error())
```
nit: "Client conn..."

easwars replied:
Same here. It doesn't provide any value to print `ctx.Err().Error()`. That would simply say something like `context deadline exceeded`. Instead change the error string to be of the form I mentioned in the other comment.

easwars replied:
And here.

arjan-bal reviewed (approved):
LGTM, please address nits and get a second approval before merging.

aranjans commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +651,191 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+	sc := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(scJSON)
+	r.InitialState(resolver.State{ServiceConfig: sc})
+
+	ccCh := make(chan *grpc.ClientConn, 1)
+	go func() {
+		cc, err := grpc.Dial(r.Scheme()+":///", grpc.WithResolvers(r), grpc.WithTransportCredentials(insecure.NewCredentials()))
+		if err != nil {
+			return
+		}
+		ccCh <- cc
+	}()
```
Yeah I had discussion for the same with Arjan [here](https://github.com/grpc/grpc-go/pull/7412#discussion_r1680852679). Let me know if you still have any questions.

easwars replied:
What I'm saying is that the same thing can be done without having to spawn a goroutine. Instead of calling `	r.InitialState(resolver.State{ServiceConfig: sc})` before the call to `grpc.Dial`, I would do the following:

- Generate the service config just like how it is being done today, at line 779.
- Then, create a grpc channel. No need of spawning a goroutine.
- Then call `r.UpdateState(resolver.State{ServiceConfig: sc})`
- Then block on a read from the `clientConnUpdateDone`

aranjans replied:
I tried removing the go routine and instead define the client just before calling r.InitialState(), but that overall call was getting stuck somewhere and it wasn't able to initialize the resolver state. I am debugging this issue, so keeping this thread open. I have resolved all the other threads.

aranjans commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +651,191 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+	cc := <-ccCh
+	defer cc.Close()
+
+	makeTestRPCAndVerifyError(ctx, t, cc, codes.Unavailable, nil)
+	t.Logf("Verifying if RPC failed when listener is stopped.")
```
Ummm..... Actually `makeTestRPCAndVerifyError` tries to make an RPC call with no metadata and in the callback defined [here](https://github.com/grpc/grpc-go/pull/7412/files#diff-f9b6e8eebb2ad5ee87acfbac326ff4f3a76bb50d25ff66559192b6e3f148c55bR737), this will throw the error as no metadata will match. And this will cause the control channel to throw the error. And that will cause the item to get into backoff. Please correct me if I am wrong. 

Also when i comment out this RPC call at line 803, the test fails which imo was expected as there won't be any item with valid backoff timer in that case.

easwars replied:
Ah yes, you are right. Could you please add a comment for this.

aranjans replied:
Added now.

aranjans commented on balancer/rls/balancer_test.go:
```diff
@@ -650,6 +651,191 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	select {
+	case <-clientConnUpdateDone:
+	case <-ctx.Done():
+		t.Fatalf("error waiting for the client conn update on initial configuration udpate: %v", ctx.Err().Error())
+	}
```
`clientConnUpdateDone` is a blocking channel and `clientConnUpdateHook` will be called twice for init/updating RLS lb config. So we'll have to wait for `clientConnUpdateDone` twice, I've tested this. WDYT?

easwars commented on balancer/rls/balancer.go:
```diff
@@ -297,13 +298,18 @@ func (b *rlsBalancer) UpdateClientConnState(ccs balancer.ClientConnState) error
 		// `stateMu` if we are to hold both locks at the same time.
 		b.cacheMu.Lock()
-		b.dataCache.resize(newCfg.cacheSizeBytes)
+		evicted := b.dataCache.resize(newCfg.cacheSizeBytes)
+		if evicted {
```
I don't understand. What does the thumbs up mean? I don't see any change being made here to address my comment.

aranjans replied:
I had not pushed the changes earlier that's why i just ack'ed the comments and did not change the assignees. I just had few questions like https://github.com/grpc/grpc-go/pull/7412#discussion_r1699923239.

aranjans commented:
@easwars I had not pushed the changes that's why i just ack'ed the comments and did not change the assignees. I just had few questions like [here](https://github.com/grpc/grpc-go/pull/7412#discussion_r1699923239).

aranjans commented on balancer/rls/balancer.go:
```diff
@@ -77,6 +77,7 @@ var (
 	clientConnUpdateHook = func() {}
 	dataCachePurgeHook   = func() {}
 	resetBackoffHook     = func() {}
+	newPickerGenerated   = func() {}
```
Updated

aranjans commented:
@easwars i have made changes in reference to all your comments except [this](https://github.com/grpc/grpc-go/pull/7412#discussion_r1702387695) one(I tried removing the go routine and instead define the client just before calling r.InitialState(), but that overall call was getting stuck somewhere and it wasn't able to initialize the resolver state. I am debugging this issue). Otherwise i feel this PR is ready to be reviewed.

easwars commented on balancer/rls/balancer.go:
```diff
@@ -297,13 +298,17 @@ func (b *rlsBalancer) UpdateClientConnState(ccs balancer.ClientConnState) error
 		b.cacheMu.Lock()
-		b.dataCache.resize(newCfg.cacheSizeBytes)
+		if b.dataCache.resize(newCfg.cacheSizeBytes) {
+			b.sendNewPickerLocked()
```
I think we need to think about this a little more. The fact that a configuration update is leading to multiple picker updates is not ideal. It is a little tricky here because of the order of locks.

aranjans replied:
Hi Easwar, based on our discussions offline I have updated the PR and ensured that we are updating picker only once.

easwars commented on balancer/rls/balancer.go:
```diff
@@ -329,6 +324,14 @@ func (b *rlsBalancer) UpdateClientConnState(ccs balancer.ClientConnState) error
 		b.dataCache.resize(newCfg.cacheSizeBytes)
 		b.cacheMu.Unlock()
 	}
+	b.stateMu.Lock()
+	// Enqueue an event which will notify us when the above update has been
+	// propagated to all child policies, and the child policies have all
+	// processed their updates, and we have sent a picker update.
+	done := make(chan struct{})
+	b.updateCh.Put(resumePickerUpdates{done: done})
+	b.stateMu.Unlock()
+	<-done
```
On second thoughts, it seems like we don't even have to acquire `stateMu` here. We are not touching any state guarded by `stateMu` here. We are only pushing a message on the `updateCh`. What do you think?

easwars commented on balancer/rls/balancer_test.go:
```diff
@@ -652,6 +652,192 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+	r := manual.NewBuilderWithScheme("rls-e2e")
+	headers := `
+    [
+        {
+            "key": "k1",
+            "names": [
+                "n1"
+            ]
+        },
+        {
+            "key": "k2",
+            "names": [
+                "n2"
+            ]
+        },
+        {
+            "key": "k3",
+            "names": [
+                "n3"
+            ]
+        }
```
Based on my earlier comment, we only need two entries here.

aranjans replied:
Ah this got here while resolving merge conflicts with master.

easwars commented on balancer/rls/balancer_test.go:
```diff
@@ -652,6 +652,192 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+	r.UpdateState(resolver.State{ServiceConfig: sc1})
+	<-clientConnUpdateDone
+	// Stop the listener
+	lis.Stop()
```
Why is this required?

easwars commented on balancer/rls/balancer_test.go:
```diff
@@ -652,6 +652,192 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	makeTestRPCAndVerifyError(ctx, t, cc, codes.Unavailable, nil)
+	t.Logf("Verifying if RPC failed when listener is stopped.")
```
This comment is not accurate.

aranjans replied:
Same this also got here while resolving conflict.

easwars commented on balancer/rls/balancer_test.go:
```diff
@@ -652,6 +652,192 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+// Test that when a data cache entry is evicted due to config change
+// in cache size, the picker is updated accordingly.
+func (s) TestPickerUpdateOnDataCacheSizeDecrease(t *testing.T) {
+	// Create a restartable listener which can close existing connections.
+	l, err := testutils.LocalTCPListener()
+	if err != nil {
+		t.Fatalf("net.Listen() failed: %v", err)
+	}
+	lis := testutils.NewRestartableListener(l)
```
We don't need a restartable listener for this test. Is there a reason to use this? Instead, you could simply pass `nil` to: `	rlsServer, rlsReqCh := rlstest.SetupFakeRLSServer(t, nil)`

aranjans replied:
Yeah my bad it was removed in earlier commits, this got back while resolving merge conflicts. Updated it now.

easwars commented on balancer/rls/balancer_test.go:
```diff
@@ -652,6 +652,192 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+
+	// Start a couple of test backends, and set up the fake RLS server to return
+	// these as targets in the RLS response, based on request keys.
+	_, backendAddress1 := startBackend(t)
```
There is no need for this backend. We only need two backends.

So, ideally your test should:
- Start two backends
- In `rlsServer.SetResponseCallback`
  - if `k1` == `v1`, return `backend1`
  - if `k2` == `v2`, return `backend2`
  - else, return error "no keys in request metadata"
- Make an RPC with no metadata. This should fail. But this will result in a data cache entry which will go into backoff
- Make an RPC with metadata (`k1=v1`). This should succeed and create a data cache entry. 
- Make an RPC with metadata (`k2=v2`). This should succeed and create a data cache entry. 
- Reduce the size of the data cache to `2`. This should evict the entry in backoff.
- Ensure that a single picker update is sent in this case.

Your test is already doing most of the above, but not all. So, please fix it a little.

aranjans replied:
Yeah this makes sense, my test was already doing earlier as you see [here](https://github.com/grpc/grpc-go/pull/7412/commits/8166fd4858bf274f2f740e37414f16563c77f9eb#diff-f9b6e8eebb2ad5ee87acfbac326ff4f3a76bb50d25ff66559192b6e3f148c55bL730). But while resolving merge conflicts, third backend got added back, apology for this. I have updated the PR now.

easwars commented on balancer/rls/balancer.go:
```diff
@@ -338,10 +331,22 @@ func (b *rlsBalancer) UpdateClientConnState(ccs balancer.ClientConnState) error
+		// specified size. If we do evict an entry with valid backoff timer,
+		// the new picker needs to be sent to the channel to re-process any
+		// RPCs queued as a result of this backoff timer.
+		//
+		// And we cannot do this operation above (where we compute the
+		// `resizeCache` boolean) because `cacheMu` needs to be grabbed before
+		// `stateMu` if we are to hold both locks at the same time.
```
This comment is now also at the top of the `if` statement. So, please delete this.

aranjans replied:
Updated ✅
